File: 34.在排序数组中查找元素的第一个和最后一个位置.go
/*
 * @lc app=leetcode.cn id=34 lang=golang
 *
 * [34] 在排序数组中查找元素的第一个和最后一个位置
 */
package leetcode

// @lc code=start
func searchRange(nums []int, target int) []int {

	return []int{leftBound(nums, target), rightBound(nums, target)}
}

func leftBound(nums []int, target int) int {
	if len(nums) == 0 {
		return -1
	}
	left, right := 0, len(nums)
	for left < right {
		mid := (left + right) / 2

		if nums[mid] < target {
			left = mid + 1
			continue
		}

		if nums[mid] > target {
			right = mid
			continue
		}

		right = mid
	}
	if left >= len(nums) || nums[left] != target {
		return -1
	}
	return left
}

func rightBound(nums []int, target int) int {
	if len(nums) == 0 {
		return -1
	}

	left, right := 0, len(nums)
	for left < right {
		mid := (left + right) / 2

		if nums[mid] < target {
			left = mid + 1
			continue
		}

		if nums[mid] > target {
			right = mid
			continue
		}

		left = mid + 1
	}

	right--
	if right < 0 || nums[right] != target {
		return -1
	}
	return right
}

// @lc code=end

File: 34.在排序数组中查找元素的第一个和最后一个位置.cpp
/*
 * @lc app=leetcode.cn id=34 lang=cpp
 *
 * [34] 在排序数组中查找元素的第一个和最后一个位置
 */
#include <vector>
using namespace std;
// @lc code=start
class Solution {
public:
  vector<int> searchRange(vector<int> &nums, int target) {
    int head = 0, tail = nums.size();

    while (head < tail) {
      int i = (tail + head) / 2;

      if (nums[i] < target) {
        head = i + 1;
        continue;
      }
      if (nums[i] > target) {
        tail = i;
        continue;
      }

      int left = i, right = i;
      while (left >= 0 && nums[left] == target) {
        left--;
      }
      while (right < nums.size() && nums[right] == target) {
        right++;
      }
      return vector<int>{++left, --right};
    }

    return vector<int>{-1, -1};
  }
};
// @lc code=end
